src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -736,6 +523,107 @@ namespace cublog
 					   definition, avg_perf_stat_results.data (), value_count);
   }
 
+  void
+  redo_parallel::set_outer_not_applied_log_lsa (const log_lsa &a_log_lsa)
+  {
+    assert (m_do_monitor_minimum_log_lsa);
+
+    m_outer_not_applied_log_lsa.store (a_log_lsa);
+  }
+
+  log_lsa
+  redo_parallel::get_calculated_minimum_not_applied_log_lsa ()
+  {
+    assert (m_do_monitor_minimum_log_lsa);
+
+    std::scoped_lock<std::mutex> slock { m_calculate_minimum_not_applied_log_lsa_mtx };
+    return m_calculated_minimum_not_applied_log_lsa;
+  }
+
+  log_lsa
+  redo_parallel::wait_past_target_lsa (const log_lsa &a_target_lsa)
+  {
+    assert (m_do_monitor_minimum_log_lsa);
+    assert (a_target_lsa != MAX_LSA);
+    assert (a_target_lsa != NULL_LSA);
+
+    log_lsa calculated_minimum_not_applied_log_lsa { MAX_LSA };
+
+    {
+      // avoid gratuitously notifying the calculating internal thread if the condition is already satisfied
+      std::scoped_lock<std::mutex> slock { m_calculate_minimum_not_applied_log_lsa_mtx };
+      calculated_minimum_not_applied_log_lsa = m_calculated_minimum_not_applied_log_lsa;
+    }
+
+    if (calculated_minimum_not_applied_log_lsa <= a_target_lsa)
+      {
+	// there is only one thread doing this calculation
+	m_calculate_minimum_not_applied_log_lsa_cv.notify_one ();
+
+	{
+	  std::unique_lock<std::mutex> ulock { m_calculate_minimum_not_applied_log_lsa_mtx };
+	  m_calculate_minimum_not_applied_log_lsa_cv.wait (ulock,
+	      [this, &calculated_minimum_not_applied_log_lsa, &a_target_lsa] ()
+	  {
+	    calculated_minimum_not_applied_log_lsa = m_calculated_minimum_not_applied_log_lsa;
+	    assert (!m_adding_finished.load ());
+	    return (calculated_minimum_not_applied_log_lsa == MAX_LSA)
+		   || (calculated_minimum_not_applied_log_lsa > a_target_lsa);
+	  });
+	}
+      }
+
+    return calculated_minimum_not_applied_log_lsa;
+  }
+
+  log_lsa
+  redo_parallel::calculate_minimum_log_lsa ()
+  {
+    assert (m_do_monitor_minimum_log_lsa);
+
+    log_lsa minimum_not_applied_log_lsa { m_outer_not_applied_log_lsa.load () };
```

Comment:
Add a comment explaining that the calculus has to start with the LSA of main thread (outer).

Reply:
Add assert is is not MAX_LSA.
```suggestion
    log_lsa minimum_not_applied_log_lsa { m_outer_not_applied_log_lsa.load () };
    assert (minimum_not_applied_log_lsa < MAX_LSA);
```

Reply:
Added an initial `log_lsa` as a ctor parameter and asserted prerequisite in the ctor.